### ISSUE TITLE
Client side exception thrown in AbstactMprUIContentConnector

### ISSUE THREAD

**Issue by TatuLund** [vaadin/multiplatform-runtime#56]
For me this exception translates to getAppId() returning null here
https://github.com/vaadin/multiplatform-runtime-internal/blob/master/mpr-core/src/main/java/com/vaadin/mpr/core/client/AbstractMprUIContentConnector.java#L128

We managed to get rid of this issue by refactoring some code we had in attach (server side) of a view to be executed later and calling super.attach() before doing anything else. I think this is some sort of potential timing problem.

After the refactoring of the code new issue https://github.com/vaadin/multiplatform-runtime/issues/54 emerged. These two may be related.

See full stack trace

```
SEVERE: Error sending hierarchy change eventscom.google.gwt.event.shared.UmbrellaException: Exception caught: (TypeError) : Cannot read property 'com_vaadin_mpr_core_client_ComponentSettings_appId' of null
    at Unknown.java_lang_Throwable_$fillInStackTrace__Ljava_lang_Throwable_2Ljava_lang_Throwable_2(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.java_lang_Throwable_Throwable__Ljava_lang_String_2Ljava_lang_Throwable_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.java_lang_RuntimeException_RuntimeException__Ljava_lang_String_2Ljava_lang_Throwable_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_google_web_bindery_event_shared_UmbrellaException_UmbrellaException__Ljava_util_Set_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.new com_google_gwt_event_shared_UmbrellaException_UmbrellaException__Ljava_util_Set_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_google_gwt_event_shared_HandlerManager_$fireEvent__Lcom_google_gwt_event_shared_HandlerManager_2Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_ui_AbstractConnector_$fireEvent__Lcom_vaadin_client_ui_AbstractConnector_2Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_ui_AbstractConnector_fireEvent__Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler$3_$sendHierarchyChangeEvents__Lcom_vaadin_client_communication_MessageHandler$3_2Lcom_vaadin_client_JsArrayObject_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler$3_$execute__Lcom_vaadin_client_communication_MessageHandler$3_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler_$handleJSON__Lcom_vaadin_client_communication_MessageHandler_2Lcom_vaadin_client_ValueMap_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler_$handleMessage__Lcom_vaadin_client_communication_MessageHandler_2Lcom_vaadin_client_ValueMap_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_mpr_core_client_AbstractMprUIContentConnector_setResponse__Ljava_lang_String_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js)
    at Unknown.<anonymous>(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js)
    at Unknown.com_google_gwt_core_client_impl_Impl_apply__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_google_gwt_core_client_impl_Impl_entry0__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.setResponse(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
Caused by: com.google.gwt.core.client.JavaScriptException: (TypeError) : Cannot read property 'com_vaadin_mpr_core_client_ComponentSettings_appId' of null
    at Unknown.com_vaadin_mpr_core_client_AbstractMprUIContentConnector_onConnectorHierarchyChange__Lcom_vaadin_client_ConnectorHierarchyChangeEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js)
    at Unknown.com_vaadin_client_ConnectorHierarchyChangeEvent_dispatch__Lcom_google_gwt_event_shared_EventHandler_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_google_web_bindery_event_shared_SimpleEventBus_$doFire__Lcom_google_web_bindery_event_shared_SimpleEventBus_2Lcom_google_web_bindery_event_shared_Event_2Ljava_lang_Object_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_google_gwt_event_shared_HandlerManager_$fireEvent__Lcom_google_gwt_event_shared_HandlerManager_2Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_ui_AbstractConnector_$fireEvent__Lcom_vaadin_client_ui_AbstractConnector_2Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_ui_AbstractConnector_fireEvent__Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler$3_$sendHierarchyChangeEvents__Lcom_vaadin_client_communication_MessageHandler$3_2Lcom_vaadin_client_JsArrayObject_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler$3_$execute__Lcom_vaadin_client_communication_MessageHandler$3_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler_$handleJSON__Lcom_vaadin_client_communication_MessageHandler_2Lcom_vaadin_client_ValueMap_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_client_communication_MessageHandler_$handleMessage__Lcom_vaadin_client_communication_MessageHandler_2Lcom_vaadin_client_ValueMap_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_vaadin_mpr_core_client_AbstractMprUIContentConnector_setResponse__Ljava_lang_String_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js)
    at Unknown.<anonymous>(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js)
    at Unknown.com_google_gwt_core_client_impl_Impl_apply__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.com_google_gwt_core_client_impl_Impl_entry0__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
    at Unknown.setResponse(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js)
 
com_google_gwt_logging_client_ConsoleLogHandler_publish__Ljava_util_logging_LogRecord_2V @ com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js:8857
```

**Comment by TatuLund** [comment]
One possible explanation for this could be that PopupExtension, especially code here, which tweaks the UI

https://github.com/parttio/PopupExtension/blob/master/popupextension-addon/src/main/java/org/vaadin/addons/popupextension/PopupExtension.java#L95

**Comment by denis-anisimov** [comment]
Does it happen if you remove `PopupExtension`  ?

**Comment by TatuLund** [comment]
We have now version running since in our test environment without PopupExtension and at the moment it looks like https://github.com/vaadin/multiplatform-runtime/issues/54 does not occur.

**Comment by TatuLund** [comment]
We have now spotted another client side exception

Caused by: com.google.gwt.core.client.JavaScriptException: (TypeError) : Cannot set property '__listener' of undefinedCaused by: com.google.gwt.core.client.JavaScriptException: (TypeError) : Cannot set property '__listener' of undefined at Unknown.com_google_gwt_user_client_impl_DOMImpl_setEventListener__Lcom_google_gwt_dom_client_Element_2Lcom_google_gwt_user_client_EventListener_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_google_gwt_user_client_ui_Widget_$onAttach__Lcom_google_gwt_user_client_ui_Widget_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.new com_vaadin_mpr_core_client_AbstractMprUIContentConnector$2_AbstractMprUIContentConnector$2__Lcom_vaadin_mpr_core_client_AbstractMprUIContentConnector_2Lcom_google_gwt_dom_client_Element_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js) at Unknown.com_vaadin_mpr_core_client_AbstractMprUIContentConnector_$attachWidget__Lcom_vaadin_mpr_core_client_AbstractMprUIContentConnector_2Lcom_google_gwt_user_client_ui_Widget_2Ljava_lang_String_2IV(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js) at Unknown.com_vaadin_mpr_core_client_AbstractMprUIContentConnector_onConnectorHierarchyChange__Lcom_vaadin_client_ConnectorHierarchyChangeEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js) at Unknown.com_vaadin_client_ConnectorHierarchyChangeEvent_dispatch__Lcom_google_gwt_event_shared_EventHandler_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_google_web_bindery_event_shared_SimpleEventBus_$doFire__Lcom_google_web_bindery_event_shared_SimpleEventBus_2Lcom_google_web_bindery_event_shared_Event_2Ljava_lang_Object_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_google_gwt_event_shared_HandlerManager_$fireEvent__Lcom_google_gwt_event_shared_HandlerManager_2Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_vaadin_client_ui_AbstractConnector_$fireEvent__Lcom_vaadin_client_ui_AbstractConnector_2Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_vaadin_client_ui_AbstractConnector_fireEvent__Lcom_google_gwt_event_shared_GwtEvent_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_vaadin_client_communication_MessageHandler$3_$sendHierarchyChangeEvents__Lcom_vaadin_client_communication_MessageHandler$3_2Lcom_vaadin_client_JsArrayObject_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_vaadin_client_communication_MessageHandler$3_$execute__Lcom_vaadin_client_communication_MessageHandler$3_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_vaadin_client_communication_MessageHandler_$handleJSON__Lcom_vaadin_client_communication_MessageHandler_2Lcom_vaadin_client_ValueMap_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_vaadin_client_communication_MessageHandler_$handleMessage__Lcom_vaadin_client_communication_MessageHandler_2Lcom_vaadin_client_ValueMap_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_vaadin_mpr_core_client_AbstractMprUIContentConnector_setResponse__Ljava_lang_String_2V(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js) at Unknown.<anonymous>(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-11.js) at Unknown.com_google_gwt_core_client_impl_Impl_apply__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-0.js) at Unknown.com_google_gwt_core_client_impl_Impl_entry0__Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2Ljava_lang_Object_2(com.kemira.ep.frontend.ui.desktop.DesktopWidgetSet-

**Comment by TatuLund** [comment]
It is also possible that this client side error from Flow client side is related as it indicates that https://github.com/vaadin/multiplatform-runtime-internal/blob/501c0cf1f8c59a9d6d0c828636f0ceb5b4383433/mpr-core/src/main/java/com/vaadin/mpr/core/AbstractLegacyEmbed.java#L117 does not work as https://github.com/vaadin/multiplatform-runtime-internal/blob/772c95933f716c6c6edd4463ff97899110a7419e/mpr-core/src/main/java/com/vaadin/mpr/core/client/AbstractMprUIContentConnector.java#L73 is probably not completed before that.

Uncaught TypeError: $0.firstElementChild.setResponse is not a function
    at Object.eval (eval at Pt (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:986), <anonymous>:3:22)
    at Pt (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:986)
    at Ot (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:943)
    at Mt (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:567)
    at sr (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:480)
    at Qr.Rr [as _] (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:994)
    at MA (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:909)
    at ur (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:989)
    at Kr.Lr [as D] (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:994)
    at bk (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:410)
    at mr (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:990)
    at nr (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:974)
    at gp.mp [as pb] (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:994)
    at Object.<anonymous> (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:961)
    at sb (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:426)
    at vb (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:861)
    at Object.onMessage (client-EFB6081741306D23AA1BBFA6CFAA8504.cache.js:606)
    at _f (vaadinPush.js?v=2.1.2:2765)
    at _invokeFunction (vaadinPush.js?v=2.1.2:2754)
    at _invokeCallback (vaadinPush.js?v=2.1.2:2884)
    at WebSocket._websocket.onmessage (vaadinPush.js?v=2.1.2:1457)

**Comment by pleku** [comment]
My bet is that this is resolved by the fix for #72. In case it is not, please comment here or open a new issue.
